velox/vector/tests/VectorSaverTest.cpp

```diff
@@ -0,0 +1,431 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/vector/VectorSaver.h"
+#include <fstream>
+#include "velox/exec/tests/utils/TempDirectoryPath.h"
+#include "velox/exec/tests/utils/TempFilePath.h"
+#include "velox/vector/fuzzer/VectorFuzzer.h"
+#include "velox/vector/tests/utils/VectorTestBase.h"
+
+namespace facebook::velox::test {
+
+class VectorSaverTest : public testing::Test, public VectorTestBase {
+ protected:
+  void assertEqualEncodings(
+      const VectorPtr& expected,
+      const VectorPtr& actual) {
+    ASSERT_EQ(expected->encoding(), actual->encoding());
+    ASSERT_EQ(*expected->type(), *actual->type());
+    assertEqualVectors(expected, actual);
+
+    switch (expected->encoding()) {
+      case VectorEncoding::Simple::CONSTANT:
+      case VectorEncoding::Simple::DICTIONARY:
+        if (expected->valueVector()) {
+          ASSERT_TRUE(actual->valueVector() != nullptr);
+          assertEqualEncodings(expected->valueVector(), actual->valueVector());
+        }
+        break;
+      default:
+          // Do nothing.
+          ;
+    }
+  }
+
+  VectorFuzzer::Options fuzzerOptions() {
+    VectorFuzzer::Options opts;
+    opts.vectorSize = 128;
+    opts.stringVariableLength = true;
+    opts.stringLength = 64;
+    return opts;
+  }
+
+  void testRoundTrip(
+      VectorFuzzer::Options options,
+      int32_t seed,
+      const TypePtr& type) {
+    VectorFuzzer fuzzer(options, pool(), seed);
+    testRoundTrip(fuzzer.fuzzFlat(type));
+  }
+
+  void testRoundTrip(const VectorPtr& vector) {
+    auto path = exec::test::TempFilePath::create();
+
+    std::ofstream outputFile(path->path, std::ofstream::binary);
+    saveVector(*vector, outputFile);
+    outputFile.close();
+
+    std::ifstream inputFile(path->path, std::ifstream::binary);
+    auto copy = restoreVector(inputFile, pool());
+    inputFile.close();
+
+    // Verify encodings and data recursively.
+    assertEqualEncodings(vector, copy);
+  }
+
+  void testRoundTrip(const TypePtr& type) {
+    auto path = exec::test::TempFilePath::create();
+
+    std::ofstream outputFile(path->path, std::ofstream::binary);
+    saveType(type, outputFile);
+    outputFile.close();
+
+    std::ifstream inputFile(path->path, std::ifstream::binary);
+    auto copy = restoreType(inputFile);
+    inputFile.close();
+
+    ASSERT_EQ(*type, *copy)
+        << "Expected: " << type->toString() << ". Got: " << copy->toString();
+  }
+
+  template <typename T>
+  void testFlatIntegers() {
+    // No nulls.
+    testRoundTrip(makeFlatVector<T>({1, 2, 3, 4, 5}));
```

Comment:
why not using the other testRoundTrip() specialization and let the fuzzer generate random input, to increase entropy?

Reply:
I'm still learning how to use the fuzzer and I find it easier to debug tests that are simply defined.